`make` fails on 18.04 with gcc 7.3.0 + LTO

Hi, I'm hitting some linker warning/errors when going through the `Building` steps on 18.04, I've tried with latest release, master, and v2 branches.

This looks to me like some sort of conflict with the included c++ files and then benchmark code but I'm not sure, any help would be appreciated. I've attached the full gist below, here is a snippet.

```
$  cmake .. -DCMAKE_BUILD_TYPE=RELEASE -DBENCHMARK_ENABLE_LTO=true -DBENCHMARK_DOWNLOAD_DEPENDENCIES=ON
...
-- Build files have been written to: /benchmark/build

$ make
...
[ 33%] Linking CXX executable cxx03_test
/usr/include/c++/7/bits/stl_tree.h:677:16: note: type name 'std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false>' should match type name 'std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>'
         struct _Rb_tree_impl
                ^
/usr/include/c++/7/bits/stl_map.h:142:41: note: type 'struct _Rep_type' should match type 'struct _Rep_type'
          key_compare, _Pair_alloc_type> _Rep_type;
                                         ^
/benchmark/include/benchmark/benchmark.h:404:40: note: type 'struct UserCounters' should match type 'struct UserCounters'
 typedef std::map<std::string, Counter> UserCounters;
                                        ^
/benchmark/include/benchmark/benchmark.h:765:14: note: type 'void Function (struct State &)' should match type 'void Function (struct State &)'
 typedef void(Function)(State&);
              ^
/benchmark/include/benchmark/benchmark.h:669:8: error: type of 'StartKeepRunning' does not match original declaration [-Werror=lto-type-mismatch]
   void StartKeepRunning();
        ^
/benchmark/src/benchmark.cc:404:6: note: implicit this pointer type mismatch
 void State::StartKeepRunning() {
      ^
/benchmark/include/benchmark/benchmark.h:448:7: note: type 'struct State' itself violates the C++ One Definition Rule
 class State {
       ^
/benchmark/src/benchmark.cc:404:6: note: 'StartKeepRunning' was previously declared here
 void State::StartKeepRunning() {
      ^
/benchmark/src/benchmark.cc:404:6: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:673:8: error: type of 'FinishKeepRunning' does not match original declaration [-Werror=lto-type-mismatch]
   void FinishKeepRunning();
        ^
/benchmark/src/benchmark.cc:412:6: note: implicit this pointer type mismatch
 void State::FinishKeepRunning() {
      ^
/benchmark/include/benchmark/benchmark.h:448:7: note: type 'struct State' itself violates the C++ One Definition Rule
 class State {
       ^
/benchmark/src/benchmark.cc:412:6: note: 'FinishKeepRunning' was previously declared here
 void State::FinishKeepRunning() {
      ^
/benchmark/src/benchmark.cc:412:6: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:984:16: error: type of 'Run' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void Run(State& st);
                ^
/benchmark/src/benchmark_register.cc:453:6: note: implicit this pointer type mismatch
 void FunctionBenchmark::Run(State& st) { func_(st); }
      ^
/benchmark/include/benchmark/benchmark.h:979:7: note: type 'struct FunctionBenchmark' itself violates the C++ One Definition Rule
 class FunctionBenchmark : public Benchmark {
       ^
/benchmark/src/benchmark_register.cc:453:6: note: 'Run' was previously declared here
 void FunctionBenchmark::Run(State& st) { func_(st); }
      ^
/benchmark/src/benchmark_register.cc:453:6: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:1057:16: error: type of 'TearDown' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void TearDown(State& st) { TearDown(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:1057:16: note: type mismatch in parameter 1
   virtual void TearDown(State& st) { TearDown(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:448:7: note: type 'struct State' itself violates the C++ One Definition Rule
 class State {
       ^
/benchmark/include/benchmark/benchmark.h:1057:16: note: 'TearDown' was previously declared here
   virtual void TearDown(State& st) { TearDown(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:1057:16: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:1056:16: error: type of 'SetUp' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void SetUp(State& st) { SetUp(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:1056:16: note: type mismatch in parameter 1
   virtual void SetUp(State& st) { SetUp(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:448:7: note: type 'struct State' itself violates the C++ One Definition Rule
 class State {
       ^
/benchmark/include/benchmark/benchmark.h:1056:16: note: 'SetUp' was previously declared here
   virtual void SetUp(State& st) { SetUp(const_cast<const State&>(st)); }
                ^
/benchmark/include/benchmark/benchmark.h:1056:16: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:1054:16: error: type of 'TearDown' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void TearDown(const State&) {}
                ^
/benchmark/include/benchmark/benchmark.h:1054:16: note: 'TearDown' was previously declared here
   virtual void TearDown(const State&) {}
                ^
/benchmark/include/benchmark/benchmark.h:1054:16: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:1053:16: error: type of 'SetUp' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void SetUp(const State&) {}
                ^
/benchmark/include/benchmark/benchmark.h:1053:16: note: 'SetUp' was previously declared here
   virtual void SetUp(const State&) {}
                ^
/benchmark/include/benchmark/benchmark.h:1053:16: note: code may be misoptimized unless -fno-strict-aliasing is used
/benchmark/include/benchmark/benchmark.h:1046:16: error: type of 'Run' does not match original declaration [-Werror=lto-type-mismatch]
   virtual void Run(State& st) {
                ^
/benchmark/include/benchmark/benchmark.h:1046:16: note: type mismatch in parameter 1
   virtual void Run(State& st) {
                ^
/benchmark/include/benchmark/benchmark.h:448:7: note: type 'struct State' itself violates the C++ One Definition Rule
 class State {
       ^
/benchmark/include/benchmark/benchmark.h:1046:16: note: 'Run' was previously declared here
   virtual void Run(State& st) {
                ^
/benchmark/include/benchmark/benchmark.h:1046:16: note: code may be misoptimized unless -fno-strict-aliasing is used
lto1: all warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
compilation terminated.
/usr/bin/x86_64-linux-gnu-ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
test/CMakeFiles/cxx03_test.dir/build.make:96: recipe for target 'test/cxx03_test' failed
make[2]: *** [test/cxx03_test] Error 1
CMakeFiles/Makefile2:293: recipe for target 'test/CMakeFiles/cxx03_test.dir/all' failed
make[1]: *** [test/CMakeFiles/cxx03_test.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

Full gist here: https://gist.github.com/tophj-ibm/c46ee3af2f1168d6e9c853cabae73dc5

Reproducible by running the build steps in an 18.04 container

Versions:
```
gcc (Ubuntu 7.3.0-16ubuntu3) 7.3.0
cmake version 3.10.2
```

Same issue with gcc 9.2.1 on Debian testing amd64.

```
[ 69%] Built target gtest
/usr/include/c++/9/bits/stl_tree.h:673:9: note: type name ‘std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false>’ should match type name ‘std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>’
  673 |  struct _Rb_tree_impl
      |         ^
/usr/include/c++/9/bits/stl_map.h:100:11: note: type ‘struct _Rep_type’ itself violates the C++ One Definition Rule
  100 |     class map
      |           ^
/home/user/benchmark/include/benchmark/benchmark.h:464:7: note: type ‘struct UserCounters’ itself violates the C++ One Definition Rule
  464 | class State {
      |       ^
/home/user/benchmark/include/benchmark/benchmark.h:677:8: error: type of ‘StartKeepRunning’ does not match original declaration [-Werror=lto-type-mismatch]
  677 |   void StartKeepRunning();
      |        ^
/home/user/benchmark/src/benchmark.cc:196:6: note: ‘StartKeepRunning’ was previously declared here
  196 | void State::StartKeepRunning() {
      |      ^
/home/user/benchmark/src/benchmark.cc:196:6: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
/home/user/benchmark/include/benchmark/benchmark.h:681:8: error: type of ‘FinishKeepRunning’ does not match original declaration [-Werror=lto-type-mismatch]
  681 |   void FinishKeepRunning();
      |        ^
/home/user/benchmark/src/benchmark.cc:204:6: note: ‘FinishKeepRunning’ was previously declared here
  204 | void State::FinishKeepRunning() {
      |      ^
/home/user/benchmark/src/benchmark.cc:204:6: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
lto1: all warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
```

Same error on ubuntu 20.04:
```Shell
[  1%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark.cc.o
[  2%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark_api_internal.cc.o
[  3%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark_name.cc.o
[  4%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark_register.cc.o
[  5%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark_runner.cc.o
[  7%] Building CXX object src/CMakeFiles/benchmark.dir/colorprint.cc.o
[  8%] Building CXX object src/CMakeFiles/benchmark.dir/commandlineflags.cc.o
[  9%] Building CXX object src/CMakeFiles/benchmark.dir/complexity.cc.o
[ 10%] Building CXX object src/CMakeFiles/benchmark.dir/console_reporter.cc.o
[ 11%] Building CXX object src/CMakeFiles/benchmark.dir/counter.cc.o
[ 12%] Building CXX object src/CMakeFiles/benchmark.dir/csv_reporter.cc.o
[ 14%] Building CXX object src/CMakeFiles/benchmark.dir/json_reporter.cc.o
[ 15%] Building CXX object src/CMakeFiles/benchmark.dir/reporter.cc.o
[ 16%] Building CXX object src/CMakeFiles/benchmark.dir/sleep.cc.o
[ 17%] Building CXX object src/CMakeFiles/benchmark.dir/statistics.cc.o
[ 18%] Building CXX object src/CMakeFiles/benchmark.dir/string_util.cc.o
[ 20%] Building CXX object src/CMakeFiles/benchmark.dir/sysinfo.cc.o
[ 21%] Building CXX object src/CMakeFiles/benchmark.dir/timers.cc.o
[ 22%] Linking CXX static library libbenchmark.a
[ 22%] Built target benchmark
[ 23%] Building CXX object src/CMakeFiles/benchmark_main.dir/benchmark_main.cc.o
[ 24%] Linking CXX static library libbenchmark_main.a
[ 24%] Built target benchmark_main
[ 25%] Building CXX object test/CMakeFiles/link_main_test.dir/link_main_test.cc.o
[ 27%] Linking CXX executable link_main_test
[ 27%] Built target link_main_test
[ 28%] Building CXX object test/CMakeFiles/register_benchmark_test.dir/register_benchmark_test.cc.o
[ 29%] Linking CXX executable register_benchmark_test
[ 29%] Built target register_benchmark_test
[ 30%] Building CXX object test/CMakeFiles/diagnostics_test.dir/diagnostics_test.cc.o
[ 31%] Linking CXX executable diagnostics_test
[ 31%] Built target diagnostics_test
[ 32%] Building CXX object test/CMakeFiles/output_test_helper.dir/output_test_helper.cc.o
[ 34%] Linking CXX static library liboutput_test_helper.a
[ 34%] Built target output_test_helper
[ 35%] Building CXX object test/CMakeFiles/templated_fixture_test.dir/templated_fixture_test.cc.o
[ 36%] Linking CXX executable templated_fixture_test
[ 36%] Built target templated_fixture_test
[ 37%] Building CXX object test/CMakeFiles/display_aggregates_only_test.dir/display_aggregates_only_test.cc.o
[ 38%] Linking CXX executable display_aggregates_only_test
[ 38%] Built target display_aggregates_only_test
[ 40%] Building CXX object test/CMakeFiles/multiple_ranges_test.dir/multiple_ranges_test.cc.o
[ 41%] Linking CXX executable multiple_ranges_test
[ 41%] Built target multiple_ranges_test
[ 42%] Building CXX object test/CMakeFiles/options_test.dir/options_test.cc.o
[ 43%] Linking CXX executable options_test
[ 43%] Built target options_test
[ 44%] Building CXX object test/CMakeFiles/skip_with_error_test.dir/skip_with_error_test.cc.o
[ 45%] Linking CXX executable skip_with_error_test
[ 45%] Built target skip_with_error_test
[ 47%] Building CXX object test/CMakeFiles/map_test.dir/map_test.cc.o
[ 48%] Linking CXX executable map_test
[ 48%] Built target map_test
[ 49%] Building CXX object third_party/googletest/build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[ 50%] Linking CXX static library ../../../../lib/libgtest.a
[ 50%] Built target gtest
[ 51%] Building CXX object third_party/googletest/build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o
[ 52%] Linking CXX static library ../../../../lib/libgmock.a
[ 52%] Built target gmock
[ 54%] Building CXX object third_party/googletest/build/googlemock/CMakeFiles/gmock_main.dir/src/gmock_main.cc.o
[ 55%] Linking CXX static library ../../../../lib/libgmock_main.a
[ 55%] Built target gmock_main
[ 56%] Building CXX object test/CMakeFiles/statistics_gtest.dir/statistics_gtest.cc.o
[ 57%] Linking CXX executable statistics_gtest
[ 57%] Built target statistics_gtest
[ 58%] Building CXX object test/CMakeFiles/benchmark_test.dir/benchmark_test.cc.o
[ 60%] Linking CXX executable benchmark_test
[ 60%] Built target benchmark_test
[ 61%] Building CXX object test/CMakeFiles/complexity_test.dir/complexity_test.cc.o
[ 62%] Linking CXX executable complexity_test
[ 62%] Built target complexity_test
[ 63%] Building CXX object test/CMakeFiles/filter_test.dir/filter_test.cc.o
[ 64%] Linking CXX executable filter_test
[ 64%] Built target filter_test
[ 65%] Building CXX object test/CMakeFiles/fixture_test.dir/fixture_test.cc.o
[ 67%] Linking CXX executable fixture_test
[ 67%] Built target fixture_test
[ 68%] Building CXX object test/CMakeFiles/cxx03_test.dir/cxx03_test.cc.o
[ 69%] Linking CXX executable cxx03_test
/usr/include/c++/9/bits/stl_tree.h:673:9: note: type name ‘std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false>’ should match type name ‘std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, benchmark::Counter> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>’
  673 |  struct _Rb_tree_impl
      |         ^
/usr/include/c++/9/bits/stl_map.h:100:11: note: type ‘struct _Rep_type’ itself violates the C++ One Definition Rule
  100 |     class map
      |           ^
/home/sam/benchmark/include/benchmark/benchmark.h:464:7: note: type ‘struct UserCounters’ itself violates the C++ One Definition Rule
  464 | class State {
      |       ^
/home/sam/benchmark/include/benchmark/benchmark.h:680:8: error: type of ‘StartKeepRunning’ does not match original declaration [-Werror=lto-type-mismatch]
  680 |   void StartKeepRunning();
      |        ^
/home/sam/benchmark/src/benchmark.cc:196:6: note: ‘StartKeepRunning’ was previously declared here
  196 | void State::StartKeepRunning() {
      |      ^
/home/sam/benchmark/src/benchmark.cc:196:6: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
/home/sam/benchmark/include/benchmark/benchmark.h:684:8: error: type of ‘FinishKeepRunning’ does not match original declaration [-Werror=lto-type-mismatch]
  684 |   void FinishKeepRunning();
      |        ^
/home/sam/benchmark/src/benchmark.cc:204:6: note: ‘FinishKeepRunning’ was previously declared here
  204 | void State::FinishKeepRunning() {
      |      ^
/home/sam/benchmark/src/benchmark.cc:204:6: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
lto1: all warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[2]: *** [test/CMakeFiles/cxx03_test.dir/build.make:86: test/cxx03_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:782: test/CMakeFiles/cxx03_test.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -194,6 +194,7 @@ else()
   # Link time optimisation
   if (BENCHMARK_ENABLE_LTO)
     add_cxx_compiler_flag(-flto)
+    add_cxx_compiler_flag(-Wno-lto-type-mismatch)
     if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
       find_program(GCC_AR gcc-ar)
       if (GCC_AR)

